src/erlang_ls_tree.erl

```diff
@@ -97,53 +97,75 @@ points_of_interest(Tree) ->
+      function      -> function(Tree, Extra);
+      implicit_fun  -> implicit_fun(Tree, Extra);
+      macro         -> macro(Tree, Extra);
+      record_access -> record_access(Tree, Extra);
+      record_expr   -> record_expr(Tree, Extra);
+      _             -> []
+    end
   catch
     Class:Reason:Stacktrace ->
-      lager:warning( "Could not analyze tree: ~p:~p ~p"
-                   , [Class, Reason, Stacktrace]),
+      Position = erl_syntax:get_pos(Tree),
+      lager:warning( "Could not analyze tree - error=~p:~p pos=~p "
+                     "stacktrace=~p"
+                   , [Class, Reason, Position, Stacktrace]),
       []
   end.
 
-%% @edoc Return the list of points of interest of a specific `Type`
-%% for a given `Tree`.
--spec points_of_interest(tree(), any(), [[erlang_ls_poi:pos()]]) -> [erlang_ls_poi:poi()].
-points_of_interest(Tree, application, Extra) ->
-  case erl_syntax_lib:analyze_application(Tree) of
-    {M, {F, A}} ->
-      %% Remote call
-      [erlang_ls_poi:poi(Tree, {application, {M, F, A}}, Extra)];
-    {F, A} ->
-      case lists:member({F, A}, erlang:module_info(exports)) of
-        true ->
-          %% Call to a function from the `erlang` module
-          [erlang_ls_poi:poi(Tree, {application, {erlang, F, A}}, Extra)];
-        false ->
-          %% Local call
-          [erlang_ls_poi:poi(Tree, {application, {F, A}}, Extra)]
+-spec application(tree(), extra()) -> [erlang_ls_poi:poi()].
+application(Tree, Extra) ->
+  MFA = case erl_syntax_lib:analyze_application(Tree) of
+          %% Remote call
+          {M, {F, A}} -> {M, F, A};
+          {F, A} ->
+            case lists:member({F, A}, erlang:module_info(exports)) of
+              %% Call to a function from the `erlang` module
+              true -> {erlang, F, A};
+              %% Local call
+              false -> {F, A}
+            end;
+          A when is_integer(A) ->
+            %% If the function is not explicitly named (e.g. a variable is
+            %% used as the module qualifier or the function name), only the
+            %% arity A is returned.
+            %% In the special case where the macro `?MODULE` is used as the
+            %% module qualifier, we can consider it as a local call.
+            Operator = erl_syntax:application_operator(Tree),
+            case erl_syntax:type(Operator) of
+              module_qualifier -> application_with_variable(Operator, A);
+              _                -> undefined
+            end
+        end,
+  case MFA of
+    undefined -> [];
+    _ -> [erlang_ls_poi:poi(Tree, {application, MFA}, Extra)]
+  end.
+
+-spec application_with_variable(tree(), arity()) ->
+  {atom(), arity()} | undefined.
+application_with_variable(Operator, A) ->
+  Module   = erl_syntax:module_qualifier_argument(Operator),
+  Function = erl_syntax:module_qualifier_body(Operator),
+  case {erl_syntax:type(Module), erl_syntax:type(Function)} of
+    %% Only in the case of ?MODULE:foo we return something
```

Comment:
Nit: instead of `?MODULE:foo` I would have something like:

> The usage of the `?MODULE` macro as the module name for fully qualified calls is so common that it is worth a specific clause.

Reply:
✅ Changed